### PR TITLE
Add ModUtils class for validating mod usages

### PR DESCRIPTION
**Pull request by smoogipoo**
Useful for both osu! and server-side validation.

### CODE REVIEW

File: osu.Game/Utils/ModUtils.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        /// <summary>
+        /// Checks that all <see cref="Mod"/>s in a combination are compatible with each-other.
+        /// </summary>
+        /// <param name="combination">The <see cref="Mod"/> combination to check.</param>
+        /// <returns>Whether all <see cref="Mod"/>s in the combination are compatible with each-other.</returns>
+        public static bool CheckCompatibleSet(IEnumerable<Mod> combination)
+        {
+            var incompatibleTypes = new HashSet<Type>();
```
Comment (bdach) [inline osu.Game/Utils/ModUtils.cs:42]: This hashset is added to, but never queried. (Weird that there's no inspection on this?)

In fact I'm not even sure I understand what the two hashsets are supposed to be tracking...

File: osu.Game.Tests/Mods/ModUtilsTest.cs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+namespace osu.Game.Tests.Mods
+{
+    [TestFixture]
+    public class ModUtilsTest
+    {
+        [Test]
+        public void TestModIsCompatibleByItself()
+        {
+            var mod = new Mock<Mod>();
+            Assert.That(ModUtils.CheckCompatibleSet(new[] { mod.Object }));
```
Comment (bdach) [inline osu.Game.Tests/Mods/ModUtilsTest.cs:18]: huh, TIL that it's not required to pass `Is.True`

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Utils/ModUtils.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Utils/ModUtils.cs:42]: I was looking at this method with the intention of combining (a portion of?) it with my follow-up function in #11642, but am also unsure how to read this code. Like I don't understand the early return or what `incomingTypes` is tracking. Is there a reason you didn't write this in a similar way to what I have in my PR?

Comment (smoogipoo) [reply osu.Game/Utils/ModUtils.cs:42]: Re: never queried: It's supposed to be querying it in the second block, not sure how it passes the tests.

It's actually doing about the same thing your code is, just in a different way. Wouldn't mind replacing it. Though I think your code should read `type.IsInstanceOfType(m)`?

Comment (peppy) [reply osu.Game/Utils/ModUtils.cs:42]: 20210202 140924 (JetBrains Rider)

That's how it reads on the current version, did you mean something else?

**Comment by smoogipoo** [comment]
Should be ready to go now. Also made some changes in the dependant PR.

**Review by peppy — APPROVED** [review]

**Pull request merged**